File: 1402-count-square-submatrices-with-all-ones/count-square-submatrices-with-all-ones.java
class Solution {
    public int countSquares(int[][] matrix) {
        int result = 0;

        for (int i = 0; i < matrix.length; i++)
            for (int j = 0; j < matrix[0].length; j++)
                if (matrix[i][j] == 1) {
                    // Look at neighbors: top, left, and top-left (diagonal)
                    int top = (i == 0) ? 0 : matrix[i - 1][j];
                    int left = (j == 0) ? 0 : matrix[i][j - 1];
                    int diag = (i == 0 || j == 0) ? 0 : matrix[i - 1][j - 1];

                    // Update current cell with the size of the largest square ending here
                    matrix[i][j] += Math.min(Math.min(top, left), diag);

                    // Add to result: each cell contributes the number of squares that end at it
                    result += matrix[i][j];
                }
        
        return result;
    }
}

File: 1402-count-square-submatrices-with-all-ones/count-square-submatrices-with-all-ones.go
func countSquares(matrix [][]int) int {
	result := 0

	for i := 0; i < len(matrix); i++ {
		for j := 0; j < len(matrix[0]); j++ {
			if matrix[i][j] == 1 {
				// look at top, left, and diagonal neighbors
				top := 0
				left := 0
				diag := 0

				if i > 0 {
					top = matrix[i-1][j]
				}

				if j > 0 {
					left = matrix[i][j-1]
				}

				if i > 0 && j > 0 {
					diag = matrix[i-1][j-1]
				}

				// update DP value
				if i > 0 && j > 0 {
					matrix[i][j] += min(top, min(left, diag))
				}

				result += matrix[i][j]
			}
		}
	}

	return result
}

func min(a, b int) int {
	if a < b {
		return a
	}
    
	return b
}